File: 1-lru-cache/python/src/cache.py
from src.node import Node

class Cache:
    def __init__(self, maxSize) -> None:
        self.head = None
        self.tail = None
        self.currentSize = 0
        self.maxSize = maxSize
        self.lookupTable = {}

    def get(self, key) -> int:
        if None == key:
            return None

        if key not in self.lookupTable.keys():
            return None

        result = self.lookupTable[key]        
        if self.tail is result:
            return result.value

        if None == result.prevNode:
            self.head = result.nextNode
            self.head.prevNode = None        

        self.tail.nextNode = result
        result.nextNode = None
        result.prevNode = None
        self.tail = result
        return result.value

    def put(self, key, val) -> None:
        if None == key:
            return None

        self.currentSize += 1
        if self.currentSize > self.maxSize:
            self.lookupTable.pop(self.head.key)
            self.head = self.head.nextNode
            self.head.prevNode = None
            self.currentSize -= 1
        
        newNode = Node(key, val, self.tail, None)
        if None == self.tail:
            self.head = newNode
        else:
            self.tail.nextNode = newNode
        
        self.tail = newNode
        self.lookupTable[key] = newNode
    
    def toArray(self):        
        a = []
        current = self.head
        while None != current:
            a.append(current.value)
            current = current.nextNode
        
        return a

File: 1-lru-cache/ruby/typescript/src/cache.ts
import { Node } from './node'

class Cache {
    readonly maxSize: number
    readonly lookupTable: Map<string, Node>
    currentSize: number;    
    head: Node
    tail: Node

    constructor(maxSize: number) {
        this.tail = null
        this.head = null
        this.currentSize = 0
        this.maxSize = maxSize
        this.lookupTable = new Map<string, Node>()
    }

    get(key: string): number {
        if (null === key) {
            return null
        }
        
        if (!this.lookupTable.has(key)) {
            return null
        }
        
        const result = this.lookupTable.get(key)
        if (this.tail === result) {
            return result.value
        }

        if (null === result.prevNode) {
            this.head = result.nextNode
            this.head.prevNode = null        
        }
        
        this.tail.nextNode = result
        result.nextNode = null
        result.prevNode = this.tail
        this.tail = result
        
        return result.value
    }

    put(key: string, val: number) {
        if (null === key) {
            return null
        }

        this.currentSize += 1
        if (this.currentSize > this.maxSize) {
            this.lookupTable.delete(this.head.key)
            this.head = this.head.nextNode
            this.head.prevNode = null
            this.currentSize -= 1
        }

        const newNode = {
            key: key, 
            value: val, 
            prevNode: this.tail,
            nextNode: null
        }

        if (null === this.tail) {
            this.head = newNode
        } else {
            this.tail.nextNode = newNode
        }        
        this.tail = newNode
        this.lookupTable.set(key, newNode)
    }

    toArray() {         
        const a = []
        var current = this.head
        while (null != current) {
            a.push(current.value)            
            current = current.nextNode            
        }               
        return a
    }
}

export { Cache }
